python/lsst/ctrl/mpexec/preExecInit.py

```diff
@@ -396,3 +415,217 @@ def savePackageVersions(self, graph: QuantumGraph) -> None:
                     self.butler.put(oldPackages, datasetType, dataId)
             else:
                 self.butler.put(packages, datasetType, dataId)
+
+
+class PreExecInitLimited:
+    """Initialization of registry for QuantumGraph execution.
+
+    This class works with LimitedButler and expects that all references in
+    QuantumGraph are resolved.
```

Comment:
(not about this line, just picking a spot to start a thread)

It's a shame we now need two versions of each of `PreExecInit`, `SingleQuantumExecutor`, and task factories.  I wonder if we could consolidate a lot of code by having single implementations that:
 - are constructed with a `LimitedButler` (which may or may not be a full one);
 - attempt to cast their butler to a full one if they are given unresolved refs, and use the butler's registry to resolve them up front - first with `findDataset` (always works for execution butler, also works for inputs with regular butler), then with `DatasetIdFactory.resolveRef` if that returns `None`;
 - proceed with using `LimitedButler` interfaces only using those resolved refs.

For `SingleQuantumExecutor` we could probably get away with transforming the `Quantum` up-front by resolving all dataset refs.  It's less clear that something similar would work for pre-exec-init and task factories right now, and that's yet another argument for special pre-exec-init quanta in the future.

Right now it seems like we're not taking advantage of the fact that `Butler` _is a_ `LimitedButler` at all.  I think it's likely that this proposal would require moving some of the butler-creation responsibilities out of these objects and into `CmdLineFwk` or `MpQuantumExecutor`, but I think that might still be an overall win.

Reply:
There is indeed a bunch of duplicate code between two executors and I'm not very happy about it. I can try to refactor things a bit and move duplicated parts into a common base class used by both of those. I would want to avoid doing `isinstance` to check Butler type, I think better OO approach is to hide those differences in subclasses (and we probably still have a mock Butler class in unit tests that is not `Butler` subclass). What worries me a bit is that if we want to use `LimitedButler` API for most things, then some of the `Butler` functionality would need to be re-implemented in ctrl_mpexec classes.